src/partitioning.c

```diff
@@ -250,13 +244,57 @@ resolve_function_argtype(FunctionCallInfo fcinfo)
 		case T_CoerceViaIO:
 			argtype = ((CoerceViaIO *) node)->resulttype;
 			break;
+		case T_FuncExpr:
+			/* Argument is function, so our input is its result type */
+			argtype = ((FuncExpr *) node)->funcresulttype;
+			break;
 		default:
-			elog(ERROR, "Unsupported expression argument node type %u", nodeTag(node));
+			elog(ERROR, "unsupported expression argument node type %u", nodeTag(node));
 	}
 
 	return argtype;
 }
 
+/*
+ * Partitioning function cache.
+ *
+ * Holds type information to avoid repeated lookups. The cache is allocated on a
+ * child memory context of the context that created the associated FmgrInfo
+ * struct. For partitioning functions invoked on the insert path, this is
+ * typically the Hypertable cache's memory context. Hence, the type cache lives
+ * for the duration of the hypertable cache and can be reused across multiple
+ * invokations of the partitioning function, even across transactions.
+ *
+ * If the partitioning function is invoked outside the insert path, the FmgrInfo
+ * and its memory context has a lifetime corresponding to that invokation.
+ */
+typedef struct PartFuncCache
+{
+	Oid			argtype;
+	Oid			coerce_funcid;
+	TypeCacheEntry *tce;
+	MemoryContext mcxt;
+} PartFuncCache;
+
+static PartFuncCache *
+part_func_cache_create(Oid argtype, TypeCacheEntry *tce, Oid coerce_funcid, MemoryContext mcxt)
+{
+	MemoryContext cache_mcxt;
+	PartFuncCache *pfc;
+
+	cache_mcxt = AllocSetContextCreate(mcxt,
```

Comment:
Why are we creating a new memory context here instead of just allocating inside mcxt?

Reply:
Maybe this was over-cautious, but since the functions take `anyelement` I figured that in theory the function could be passed different argument types over the lifetime of the cache, so the `ParfFuncCache` would have to be reallocated/changed to reflect the new type. However, I don't think this ever happens in practice and the code certainly doesn't account for it right now. 

It might make sense to allocate on `mcxt` for now and rework this if needed in the future. I'll update.